File: web/apps/admin-web/src/api/savedViews.test.ts
import { describe, expect, it, vi } from "vitest";
import type { SavedViewStateV1 } from "@xingmang/ui-admin";
import type { ApiClient } from "./client";
import {
  listSavedViews,
  removeSavedView,
  setSavedView,
} from "./savedViews";

function fakeClient(body: unknown): ApiClient {
  return {
    get: vi.fn().mockResolvedValue(body),
    post: vi.fn().mockResolvedValue(body),
  } as unknown as ApiClient;
}

const state: SavedViewStateV1 = {
  schema_version: 1,
  query: "openai",
  filters: { status: "需关注" },
  sort: { column_id: "grossProfit", direction: "desc" },
  columns: { known: ["name", "status", "grossProfit"], visible: ["name", "status"] },
  density: "compact",
};

describe("SavedView API", () => {
  it("Query 只传 table_key，特殊字符由统一客户端编码", async () => {
    const client = fakeClient({ items: null });
    await expect(listSavedViews("platform.sub2api.channels/测试", {}, client)).resolves.toEqual([]);
    expect(client.get).toHaveBeenCalledWith("/api/v1/ui/saved-views", {
      searchParams: { table_key: "platform.sub2api.channels/测试" },
    });
  });

  it("set 走 ui.saved_view.set@1，v1 状态展平成既有 Schema 原语", async () => {
    const client = fakeClient({ action_run_id: "run-set-1", result: { id: "view-1" } });
    const run = await setSavedView(
      { tableKey: "platform.sub2api.channels", name: "需关注", state },
      {},
      client,
    );
    expect(run.runId).toBe("run-set-1");
    const call = (client.post as ReturnType<typeof vi.fn>).mock.calls[0];
    expect(call?.[0]).toBe("/api/v1/actions/ui.saved_view.set/versions/1/execute");
    expect(call?.[1]).toEqual({
      params: {
        table_key: "platform.sub2api.channels",
        name: "需关注",
        schema_version: 1,
        query: "openai",
        filters_json: `{"status":"需关注"}`,
        sort_column: "grossProfit",
        sort_direction: "desc",
        known_columns: ["name", "status", "grossProfit"],
        visible_columns: ["name", "status"],
        density: "compact",
      },
    });
    const params = (call?.[1] as { params: Record<string, unknown> }).params;
    for (const forbidden of ["owner_issuer", "owner_subject", "identity_zone", "environment"]) {
      expect(params).not.toHaveProperty(forbidden);
    }
  });

  it("无排序时两个排序字段都传空串，不能出现半对", async () => {
    const client = fakeClient({ action_run_id: "run-set-2" });
    await setSavedView(
      { tableKey: "global.audit.events", name: "默认列", state: { ...state, sort: null } },
      {},
      client,
    );
    const body = (client.post as ReturnType<typeof vi.fn>).mock.calls[0]?.[1] as {
      params: Record<string, unknown>;
    };
    expect(body.params.sort_column).toBe("");
    expect(body.params.sort_direction).toBe("");
  });

  it("remove 走 ui.saved_view.remove@1，只传 UUID", async () => {
    const client = fakeClient({ action_run_id: "run-remove-1" });
    const run = await removeSavedView("view-1", {}, client);
    expect(run.runId).toBe("run-remove-1");
    const call = (client.post as ReturnType<typeof vi.fn>).mock.calls[0];
    expect(call?.[0]).toBe("/api/v1/actions/ui.saved_view.remove/versions/1/execute");
    expect(call?.[1]).toEqual({ params: { saved_view_id: "view-1" } });
  });
});

File: internal/platform/httpapi/savedviews_test.go
package httpapi

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/google/uuid"

	"github.com/xufei5620/xingmang-platform/internal/platform/principal"
	"github.com/xufei5620/xingmang-platform/internal/platform/savedviews"
)

type fakeSavedViewLister struct {
	items []savedviews.SavedView
	err   error
	owner savedviews.Owner
	table string
}

func (f *fakeSavedViewLister) List(_ context.Context, owner savedviews.Owner, table string) ([]savedviews.SavedView, error) {
	f.owner, f.table = owner, table
	return f.items, f.err
}

func savedViewPrincipal(typ principal.Type) principal.Principal {
	return principal.Principal{
		ID: "staff_alice", Type: typ, IdentityZone: "staff",
		Issuer: "https://auth.example/realms/staff", Subject: "sub-alice",
		Environment: "production", Scopes: []string{savedviews.ScopeManage},
	}
}

func savedViewItem() savedviews.SavedView {
	now := time.Date(2026, 8, 29, 1, 2, 3, 0, time.UTC)
	return savedviews.SavedView{
		ID:       uuid.MustParse("3c6d7c6f-5eec-4db4-8a23-55754aa50ceb"),
		TableKey: "platform.sub2api.channels", Name: "需关注",
		State: savedviews.StateV1{
			SchemaVersion: 1, Query: "openai", Filters: map[string]string{"status": "需关注"},
			Sort:           &savedviews.Sort{ColumnID: "grossProfit", Direction: savedviews.SortDesc},
			KnownColumns:   []string{"name", "status", "grossProfit"},
			VisibleColumns: []string{"name", "status"}, Density: savedviews.DensityCompact,
		},
		StateHash: strings.Repeat("a", 64), CreatedAt: now, UpdatedAt: now,
	}
}

func callSavedViews(t *testing.T, store SavedViewLister, rawURL string, p principal.Principal) *httptest.ResponseRecorder {
	t.Helper()
	req := httptest.NewRequest(http.MethodGet, rawURL, nil)
	req = req.WithContext(principal.WithPrincipal(req.Context(), p))
	recorder := httptest.NewRecorder()
	ListSavedViewsHandler(store).ServeHTTP(recorder, req)
	return recorder
}

func TestListSavedViewsReturnsExplicitOwnerSafeDTO(t *testing.T) {
	store := &fakeSavedViewLister{items: []savedviews.SavedView{savedViewItem()}}
	recorder := callSavedViews(t, store, "/api/v1/ui/saved-views?table_key=platform.sub2api.channels", savedViewPrincipal(principal.TypeHuman))
	if recorder.Code != http.StatusOK {
		t.Fatalf("status=%d body=%s", recorder.Code, recorder.Body.String())
	}
	if store.owner.Subject != "sub-alice" || store.owner.Environment != "production" || store.table != "platform.sub2api.channels" {
		t.Fatalf("store call = %#v / %q", store.owner, store.table)
	}
	var body map[string]any
	if err := json.Unmarshal(recorder.Body.Bytes(), &body); err != nil {
		t.Fatal(err)
	}
	encoded := recorder.Body.String()
	for _, forbidden := range []string{"owner_issuer", "owner_subject", "identity_zone", "environment", "state_hash"} {
		if strings.Contains(encoded, forbidden) {
			t.Fatalf("response leaked %s: %s", forbidden, encoded)
		}
	}
	for _, required := range []string{`"table_key"`, `"state_version"`, `"schema_version"`, `"columns"`, `"updated_at"`} {
		if !strings.Contains(encoded, required) {
			t.Fatalf("response missing %s: %s", required, encoded)
		}
	}
}

func TestListSavedViewsRejectsMissingOrClientBoundaryParams(t *testing.T) {
	for _, rawURL := range []string{
		"/api/v1/ui/saved-views",
		"/api/v1/ui/saved-views?table_key=platform.sub2api.channels&environment=staging",
		"/api/v1/ui/saved-views?table_key=platform.sub2api.channels&owner_subject=attacker",
	} {
		recorder := callSavedViews(t, &fakeSavedViewLister{}, rawURL, savedViewPrincipal(principal.TypeHuman))
		if recorder.Code != http.StatusBadRequest {
			t.Fatalf("%s status=%d body=%s", rawURL, recorder.Code, recorder.Body.String())
		}
	}
}

func TestListSavedViewsRejectsMachinePrincipal(t *testing.T) {
	for _, typ := range []principal.Type{principal.TypeService, principal.TypeAI, principal.TypeServerAgent} {
		recorder := callSavedViews(t, &fakeSavedViewLister{}, "/api/v1/ui/saved-views?table_key=global.audit.events", savedViewPrincipal(typ))
		if recorder.Code != http.StatusForbidden {
			t.Fatalf("%s status=%d body=%s", typ, recorder.Code, recorder.Body.String())
		}
	}
}

func TestSavedViewRouteRequiresDedicatedScope(t *testing.T) {
	resolver, err := NewDevHeaderResolver("development")
	if err != nil {
		t.Fatal(err)
	}
	store := &fakeSavedViewLister{items: []savedviews.SavedView{}}
	handler := NewRouter(Deps{
		Logger: discardLogger(), Service: "platform-api", Environment: "development",
		DB: fakePinger{}, Resolver: resolver, SavedViews: store,
	})
	for _, tc := range []struct {
		scopes string
		want   int
	}{
		{"registry.read,ops.read", http.StatusForbidden},
		{savedviews.ScopeManage, http.StatusOK},
	} {
		req := httptest.NewRequest(http.MethodGet, "/api/v1/ui/saved-views?table_key=global.audit.events", nil)
		devHeaders(req, tc.scopes)
		recorder := httptest.NewRecorder()
		handler.ServeHTTP(recorder, req)
		if recorder.Code != tc.want {
			t.Fatalf("scopes=%q status=%d body=%s", tc.scopes, recorder.Code, recorder.Body.String())
		}
	}
}
